tikv/raftstore/snap_builder.go

```diff
@@ -0,0 +1,231 @@
+			return err
+		}
+		b.cfFiles[defaultCFIdx].KVCount++
+		b.size += len(defaultCFKey) + len(l.Value)
+		b.kvCount++
+	}
+	b.buf = codec.EncodeCompactBytes(b.buf[:0], lockCFKey)
+	_, err := b.lockCFWriter.Write(b.buf)
+	if err != nil {
+		return err
+	}
+	b.size += len(b.buf)
+	b.buf2 = encodeLockCFValue(lockCFVal, b.buf2[:0])
+	b.buf = codec.EncodeCompactBytes(b.buf[:0], b.buf2)
+	_, err = b.lockCFWriter.Write(b.buf)
+	if err != nil {
+		return err
+	}
+	b.cfFiles[lockCFIdx].KVCount++
+	b.size += len(b.buf)
+	b.kvCount++
+
+	b.lockIterator.Next()
+	if b.lockIterator.Valid() && !b.reachEnd(b.lockIterator.Key()) {
+		b.curLockKey = b.lockIterator.Key()
+	} else {
+		b.curLockKey = nil
+	}
+	return nil
+}
+
+func (b *snapBuilder) addDBEntry() error {
+	item := b.dbIterator.Item()
+	val, err := item.Value()
+	if err != nil {
+		return err
+	}
+	meta := mvcc.DBUserMeta(item.UserMeta())
+	err = b.addSSTKey(b.curDBKey, meta.StartTS(), meta.CommitTS(), val)
+	if err != nil {
+		return err
+	}
+	err = b.addOldWrite()
+	if err != nil {
+		return err
+	}
+	b.dbIterator.Next()
+	if b.dbIterator.Valid() && !b.reachEnd(b.dbIterator.Item().Key()) {
+		b.curDBKey = b.dbIterator.Item().Key()
+	} else {
+		b.curDBKey = nil
+	}
+	return nil
+}
+
+func (b *snapBuilder) addOldWrite() error {
+	for {
+		if !b.dbOldIterator.Valid() {
+			return nil
+		}
+		item := b.dbOldIterator.Item()
+		oldKey := item.Key()
+		if !isLatestKeyAndOldKeySame(b.curDBKey, oldKey) {
+			return nil
+		}
+		_, commitTS, err := codec.DecodeUintDesc(oldKey[len(oldKey)-8:])
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		meta := mvcc.OldUserMeta(item.UserMeta())
+		val, err := item.Value()
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		err = b.addSSTKey(b.curDBKey, meta.StartTS(), commitTS, val)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		b.dbOldIterator.Next()
+	}
+}
+
+func (b *snapBuilder) addSSTKey(key []byte, startTS, commitTS uint64, val []byte) error {
+	writeCFKey := encodeRocksDBSSTKey(key, commitTS)
+	writeCFVal := new(writeCFValue)
+	if len(val) == 0 {
+		writeCFVal.writeType = byte(kvrpcpb.Op_Del)
+	} else {
+		writeCFVal.writeType = byte(kvrpcpb.Op_Put)
+	}
+	writeCFVal.startTS = startTS
+	if len(val) <= shortValueMaxLen {
+		writeCFVal.shortValue = val
+	} else {
+		defaultCFKey := encodeRocksDBSSTKey(b.curDBKey, startTS)
+		err := b.defaultCFWriter.Put(defaultCFKey, val)
+		if err != nil {
+			return err
+		}
+		b.cfFiles[defaultCFIdx].KVCount++
+		b.kvCount++
+		b.size += len(defaultCFKey) + len(val)
+	}
+	encodedWriteCFVal := encodeWriteCFValue(writeCFVal)
+	err := b.writeCFWriter.Put(writeCFKey, encodedWriteCFVal)
+	if err != nil {
+		return err
+	}
+	b.cfFiles[writeCFIdx].KVCount++
+	b.size += len(writeCFKey) + len(encodedWriteCFVal)
+	b.kvCount++
+	return nil
+}
+
+func isLatestKeyAndOldKeySame(lastestKey, oldKey []byte) bool {
+	return len(oldKey)-len(lastestKey) == 8 && bytes.Equal(lastestKey[1:], oldKey[1:len(lastestKey)])
```

Comment:
There is a potential optimization that compare the last byte after length match, because these two keys may have several common beginning bytes.

Reply:
We can do this optimization when we see it in profile.